addons/mail/controllers/search.py

```diff
@@ -0,0 +1,45 @@
+from odoo import http
+from odoo.http import request
+from odoo.osv import expression
+
+
+class MailSearchController(http.Controller):
+    @http.route('/mail/thread/search', methods=['POST'], type='json', auth='user')
+    def mail_thread_search(self, search_term, thread_model, thread_id, **kwargs):
+        domain = []
+        domain = expression.AND([
+            expression.OR([
+                [('body', 'ilike', search_term)],
+                [('tracking_value_ids.field_desc', 'ilike', search_term)],
```

Comment:
Tracking values is not accessible to anyone else than admin normally, so this won't return you anything. This model is technical anyway, we should not let people search in it, just offer some tools on it.

Reply:
And yet they are displayed for Marc Demo in the Chatter.
You are right that there is a permission issue, but we are displaying information and that is supposed to be restricted ? I don't understand...

Reply:
Hello,

Indeed, it is security model based on delegation, it is not obvious at first glance. Idea is: data is not accessible, but provided by secure entry points.

Model is not available to anyone except admin, because it is considered as a technical model, holding values that should not be accessible directly. Example: private information on employee, salary information, tokens, ... Idea is that they may contain information people should not be able to read, access, ... without any context.

Then message offers a way to display them. Currently we delegate the choice to display values of a message to message_fetch (or equivalent), that selects tracking values to display to the customer. If you look at the model, you have a ``field_groups`` computed field on tracking model that either take the field ``groups`` attribute (e.g. field limited to a user group, like HR managers for private information), either are limited to admin if there is no field (e.g. some custom data storage, never accessible to anyone else than admin).

I don't think people should be able to search on tracking values. Otherwise you have to implement the ``field_groups`` logic directly in a search (performances ?), and also it eases finding information by doing incremental searches. Some information that is not really harmful when being given only within a message context could lead to guessing more important information if people can search on it.

Finally I don't think this adds any value to a search. If I search messages on "canceled" I don't think I want to find any message linked to the canceled stage of a task. For that purpose I have standard filters in Odoo that are much more efficient, reports, ... If you want to include a partner's name, then search in the relevant model for the exact partner_id. Is there any scenario where searching on tracking values is relevant and useful, compared to the performance, security and complexity issues the code is adding ?

Reply:
Hello !

I ll see what I can do about this, I m really unhappy to have to deal with that amount of security stuff.

> Finally I don't think this adds any value to a search. If I search messages on "canceled" I don't think I want to find any message linked to the canceled stage of a task. For that purpose I have standard filters in Odoo that are much more efficient, reports, ... If you want to include a partner's name, then search in the relevant model for the exact partner_id. Is there any scenario where searching on tracking values is relevant and useful, compared to the performance, security and complexity issues the code is adding ?

*Picture this:*

I m a user, searching inside the chatter, for me, tracking value are the same has  message (chatter display them the same), I m searching for `in progress` to find when a task staged change to `in progress` and the search returns nothing.

Now, what does the user think about the search feature inside the chatter? Because I personally will think that it's a bug.

Reply:
I think this feature is actually unnecessary. Either you are a project manager, and you want better tools (reporting, analysis, ...) that is already managed in Odoo by specific tools.  Either you just want to be able to sometimes find an information that you can find using ctrl-f on your page.

Moreover for managers, searching on tracking values does not add anything functionally, that's not what a PM expects. For example on projet and lead we will add a reporting of time spent in each stage, based on tracking values. That is something that is actionnable for people: KPIs, real timing, ... integrated into the UI. That is what makes sense functionally.

As there is no task linked, no functional explanation, for me this is more a developer task than a real functional task. You want to be really precise in your control and what you return. But functionally, searching on "canceled" is something you never do. You use reporting, grouping, tasks analysis... or you juste scroll the chatter whenever you want the information.

Anyway, all in all, you have to at least respect security and privacy. You can't say "I am unhappy with security". Odoo is not a tool for developers, but something containing a lot of information that should not be leaked. It should be your main concern when dealing with tracking value actually, not being able to search on "in progress".

Cheers,

Reply:
(tde is on leave for maybe a couple weeks now so I wouldn't expect a response too soon)

I think the difference is simply that we never normally allow search on tracking values when fetching messages through the other entry points. You still have an issue despite _message_format because you leak information through the search. _message_format hides the tracking values, but only the tracking values. 

If you search the value of a selection field, and you notice you get a message that does not contain that string, you know that the message is saying this record had that value for that field at some point, and you even know when it changed. That is problematic for lots of things.

You would need to explicitly check that the tracking values you retrieve do not correspond to a field that is not accessible by the user. Like we do in _message_format. And you would only want to return a message if it either contained tracking values for fields I have access to, or if it contains the search term in some other way (like in the body). Effectively, you need 2 searches. Plus all the queries involved in checking the permissions.

Reply:
Thank you @thiryr now I understand the issue.
I added a step to filter out message with tracking_values that can't be seen by the user. Should be good now.